File: src/arrays/twoPointers/isSubsequence.js
/**
 * 392. Is Subsequence
 * https://leetcode.com/problems/is-subsequence/?envType=study-plan-v2&envId=leetcode-75
 * -------------------------------------------------------------
 * Given two strings s and t, return true if s is a subsequence of t, or false otherwise.
 *
 * A subsequence of a string is a new string that is formed from the original string by
 * deleting some (can be none) of the characters without disturbing the relative positions
 * of the remaining characters. (i.e., "ace" is a subsequence of "abcde" while "aec" is not).
 *
 * Example 1:
 *
 * Input: s = "abc", t = "ahbgdc"
 * Output: true
 *
 * Example 2:
 *
 * Input: s = "axc", t = "ahbgdc"
 * Output: false
 *
 * Constraints:
 *
 * 0 <= s.length <= 100
 * 0 <= t.length <= 104
 * s and t consist only of lowercase English letters.
 * -------------------------------------------------------------
 *
 * Runtime complexity: O(n)
 * Space complexity: O(1)
 *
 *
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
let isSubsequence = function(possibleSubString,stringOfChars) {
  let pointer1 = 0;

  for (let i=0; i<stringOfChars.length; i++) {
    if (possibleSubString[pointer1] === stringOfChars[i]) {
      pointer1++;
    }
  }

  return pointer1 === possibleSubString.length;
};

let possibleSubString = "abc";
let stringOfChars = "ahbgdc";

console.log(isSubsequence(possibleSubString,stringOfChars));

let s = "axc";
let t = "ahbgdc";
console.log(isSubsequence(s,t));

s = "a";
t = "a";
console.log(isSubsequence(s,t));

s = "";
t = "";
console.log(isSubsequence(s,t));

File: ruby/two_pointers/is_subsequence.rb
# @param {String} s
# @param {String} t
# @return {Boolean}
def is_subsequence(s, t)
  position_in_s = 0
  position_in_t = 0

  while position_in_t < t.size
    if s[position_in_s] == t[position_in_t]
      position_in_s +=1
      position_in_t +=1
    else
      position_in_t +=1
    end
  end

  position_in_s == s.size
end

s = "abc"
t = "ahbgdc"
p is_subsequence(s, t) # => true

s = "a"
t = "ba"
p is_subsequence(s, t)

s = "axc"
t = "ahbgdc"
p is_subsequence(s, t) #